### PR TITLE
Coalesce unreachable_patterns lint messages

### PR DESCRIPTION
**Pull request by mejrs**
Rather than emit a message for every unreachable pattern in a match, this PR coalesces them to one message. This is part of an effort to make mistakes like https://github.com/rust-lang/rust/issues/68433 less spammy.

For example, it combines all of these...

```rust
error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:23:9
   |
LL |         (1, 3) => {}
   |         ^^^^^^

error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:24:9
   |
LL |         (1, 4) => {}
   |         ^^^^^^

error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:25:9
   |
LL |         (2, 4) => {}
   |         ^^^^^^

error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:26:9
   |
LL |         (2 | 1, 4) => {}
   |         ^^^^^^^^^^

error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:28:9
   |
LL |         (1, 4 | 5) => {}
   | 
```

to...

```rust
error: multiple unreachable patterns
  --> $DIR/exhaustiveness-unreachable-pattern.rs:29:9
   |
LL |         (1, 3) => {}
   |         ^^^^^^^^^^^^ this arm is never executed
...
LL |         (1, 4) => {}
   |         ^^^^^^^^^^^^ this arm is never executed
LL |
LL |         (2, 4) => {}
   |         ^^^^^^^^^^^^ this arm is never executed
LL |
LL |         (2 | 1, 4) => {}
   |         ^^^^^^^^^^^^^^^^ this arm is never executed
...
LL |         (1, 4 | 5) => {}
   |         ^^^^^^^^^^^^^^^^ this arm is never executed
```

### CODE REVIEW

File: compiler/rustc_error_messages/locales/en-US/mir_build.ftl [hunk 1]
```diff
@@ -192,9 +192,12 @@ mir_build_const_param_in_pattern = const parameters cannot be referenced in patt
 
 mir_build_non_const_path = runtime values cannot be referenced in patterns
 
-mir_build_unreachable_pattern = unreachable pattern
-    .label = unreachable pattern
-    .catchall_label = matches any value
+mir_build_unreachable_patterns = one or more patterns are unreachable
```
Comment (fmease) [inline compiler/rustc_error_messages/locales/en-US/mir_build.ftl:195]:
```suggestion
mir_build_unreachable_patterns = {$amount ->
    [one] unreachable pattern
    *[other] multiple unreachable patterns
}
```

Looks a bit cleaner in my opinion. Take it or leave it though.

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.rs [hunk 2]
```diff
@@ -4,81 +4,107 @@
 fn main() {
```
Comment (fmease) [inline tests/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:4]: I am not sure what the others think but an argument can be made that every single flagged unreachable pattern is of equal importance within the diagnostic and thus each span should be a primary one (which I think can be realized with `DiagnosticBuilder::set_span` which accepts a `MultiSpan`).

File: compiler/rustc_error_messages/locales/en-US/mir_build.ftl (hunk 1 above)
Comment (mejrs) [reply compiler/rustc_error_messages/locales/en-US/mir_build.ftl]: Done

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.rs (hunk 2 above)
Comment (mejrs) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:4]: Done 👍

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs [hunk 3]
```diff
@@ -139,6 +139,7 @@ impl<'tcx, 'body> ParseCtxt<'tcx, 'body> {
     }
 
     fn parse_rvalue(&self, expr_id: ExprId) -> PResult<Rvalue<'tcx>> {
+        #![allow(unreachable_patterns)]
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/custom/parse/instruction.rs:142]: why are these needed when they weren't before?

Also please attach the attribute to the function instead of using an inner attribute

File: tests/ui/consts/const_in_pattern/issue-78057.stderr [hunk 4]
```diff
@@ -5,13 +5,13 @@ LL |         FOO => {},
...
-  --> $DIR/issue-78057.rs:14:9
+  --> $DIR/issue-78057.rs:12:9
    |
 LL |         FOO => {},
-   |         --- matches any value
+   |         ^^^ this pattern is irrefutable; subsequent arms are never executed
 LL |
 LL |         _ => {}
-   |         ^ unreachable pattern
+   |         ^^^^^^^ this arm is never executed
```
Comment (oli-obk) [inline tests/ui/consts/const_in_pattern/issue-78057.stderr:14]: this may get very verbose for longer arms, maybe keep pointing to the pattern?

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.rs [hunk 5]
```diff
@@ -131,22 +165,25 @@ fn main() {
     // https://github.com/rust-lang/rust/issues/76836
     match None {
         Some(false) => {}
-        None | Some(true
-                | false) => {} //~ ERROR unreachable
+        None | Some(true | false) => {} //~ ERROR unreachable
```
Comment (oli-obk) [inline tests/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:168]: this formatting was on purpose, to make sure the span doesn't point to the `None` or the `true`

File: tests/ui/consts/const_in_pattern/issue-78057.stderr (hunk 4 above)
Comment (oli-obk) [reply tests/ui/consts/const_in_pattern/issue-78057.stderr:14]: Looking at the other diagnostic changes I think we should just keep the previous `unreachable pattern`, considering the other message is so much more helpful now

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs (hunk 3 above)
Comment (mejrs) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs:145]: Since I changed how the spans work (not entirely sure how/why), it will emit an error like this:
```rust
error: unreachable pattern
   --> compiler\rustc_mir_build\src\build\custom\parse.rs:54:13
    |
22  | / macro_rules! parse_by_kind {
23  | |     (
24  | |         $self:ident,
25  | |         $expr_id:expr,
...   |
54  | |             _ => return Err($self.expr_error(expr_id, $expected))
    | |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this arm is never executed
55  | |         }
56  | |     }};
57  | | }
    | |_- in this expansion of `parse_by_kind!`
    |
   ::: compiler\rustc_mir_build\src\build\custom\parse\instruction.rs:146:9
    |
146 | /         parse_by_kind!(self, expr_id, _, "rvalue",
147 | |             @call("mir_discriminant", args) => self.parse_place(args[0]).map(Rvalue::Discriminant),
148 | |             @call("mir_checked", args) => {
149 | |                 parse_by_kind!(self, args[0], _, "binary op",
...   |
171 | |             _ => self.parse_operand(expr_id).map(Rvalue::Use),
    | |             ^ this pattern is irrefutable; subsequent arms are never executed
172 | |         )
    | |_________- in this macro invocation
    |
    = note: `-D unreachable-patterns` implied by `-D warnings`
```

File: tests/ui/consts/const_in_pattern/issue-78057.stderr (hunk 4 above)
Comment (mejrs) [reply tests/ui/consts/const_in_pattern/issue-78057.stderr:14]: What do you mean with "the other message"? I'm afraid I'm not sure what you are proposing.

Comment (mejrs) [reply tests/ui/consts/const_in_pattern/issue-78057.stderr:14]: Anyway: newbies tend to be more experienced with switch...case structures than pattern matching so I prefer saying "this arm is never executed" over "pattern is unreachable". It'd be kinda weird to say "this arm is never executed" and just point at the pattern rather than the pattern and the entire code block.

Comment (oli-obk) [reply tests/ui/consts/const_in_pattern/issue-78057.stderr:14]: I understand that argument, but it is orthogonal to the "make this diagnostic less spammy" change, so I would prefer to discuss it separately with wg-diagnostics, as it seems like a major divergence from existing diagnostics and we'd need to adjust other diagnostics, too.

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: please allow this lint inside the macro instead of at the call sites then.

Comment (mejrs) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: I just realized why this happens, it's because I now use the HirId of the scrutinee rather than the match arms, so I guess (the documentation doesn't say) this means the allow attributes now have to be on the match construct itself rather than individual arms.

Actually that seems really undesirable to me so unless there is a way around this that I didn't see, I'm not sure this PR is a good idea.

Comment (oli-obk) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: Oh good catch. I guess we could use the hir id of the entire match instead?

Comment (estebank) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: We need to run crater to see how this will impact the ecosystem.

File: tests/ui/consts/const_in_pattern/issue-78057.stderr (hunk 4 above)
Comment (estebank) [reply tests/ui/consts/const_in_pattern/issue-78057.stderr:14]: I'm in agreement with oli's points. We want to make the underlines as short as possible across the board. To address your concern around wording, you could say "this pattern cannot be reached, so its arm is never executed", while pointing only at the pattern.

File: compiler/rustc_mir_build/src/build/custom/parse/instruction.rs (hunk 3 above)
Comment (mejrs) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: > We need to run crater to see how this will impact the ecosystem.

From some basic searches I can tell there will be many matches that will need to shift their `allow`s. Most of them are to deal with conditionally compiling match arms.

Comment (mejrs) [reply compiler/rustc_mir_build/src/build/custom/parse/instruction.rs]: Which I really don't like, actually. I think the allows should remain on the match arms.

**Comment by mejrs** [comment]
Closing because I no longer think this is an improvement. Thanks for the reviews :)
